dbms/src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -444,6 +462,11 @@ bool ZooKeeper::exists(const std::string & path, Stat * stat_, const EventPtr &
     return existsWatch(path, stat_, callbackForEvent(watch));
 }
 
+bool ZooKeeper::exists(const std::string & path, Stat * stat, const TaskHandlePtr & watch)
```

Comment:
I don't think this class should know about TaskHandles. Can we use the generic callback interface (with TaskHandles converted to callbacks outside the ZooKeeper class)?

Reply:
This is a bit tricky. If we want a uniform implementation we need 2 implementations for that interface. One using the EventPtr and one using TaskHandlePtr. Because both functions are doing same stuff.. only signaling it's done via a different method.

What do you think ?

Reply:
We should have 2 impls for each method: one with EventPtr and one with WatchCallback (generic callback interface). Clients of BackgroundSchedulePool that want their tasks to be scheduled when a watch fires, should create WatchCallbacks themselves. This way ZooKeeper and BackgroundSchedulePool can be decoupled.

Reply:
OK noted !

Reply:
Fixed in commit a2dc16a582540c2bc07f3060ad0fe0e191e75e95

Reply:
👍